t/lib/WTSI/NPG/HTS/TreePublisherTest.pm

```diff
@@ -109,40 +110,49 @@ sub publish_tree : Test(58) {
 
 sub npg_publish_tree_pl_metadata_from_stdin : Test(3) {
   my $source_path = "${data_path}/treepublisher";
-  my @attributes;  
-  push @attributes, {attribute => 'attr1', value => 'val1', units => q[]};
-  push @attributes, {attribute => 'attr2', value => 'val2', units => q[]};
+  my @attributes = (  
+    {attribute => 'attr1', value => 'val1', units => q[]},
+    {attribute => 'attr2', value => 'val2', units => q[]}
+  );
 
   my $metadata_text = JSON->new->utf8->encode(\@attributes);
-  my $metadata_file_in = "metadata.json.in";
-  open my $md_stdio, '>', $metadata_file_in or die "Cannot open ${metadata_file_in} for test";
-  print $md_stdio "$metadata_text\n";
+  my $md_stdio = File::Temp->new(SUFFIX => ".json");
+  my $metadata_file_in = $md_stdio->filename;
+  print $md_stdio "${metadata_text}\n";
 
-  my $script_args = qq[--collection ${irods_tmp_coll} --source_directory ${source_path} -];
-  ok(system("cat ${metadata_file_in} | ${bin_path}/npg_publish_tree.pl ${script_args}") == 0, 'Script npg_publish_tree.pl with metadata in stdin');
+  my $script_args = "--collection ${irods_tmp_coll} " . 
+                      "--source_directory ${source_path} -";
+  ok(system("cat ${metadata_file_in} | " .
+            "${bin_path}/npg_publish_tree.pl ${script_args}") == 0,
+      'Script npg_publish_tree.pl with metadata in STDIN');
 
   my $imeta_output = `imeta ls -C ${irods_tmp_coll}`;
```

Comment:
It's preferable to use the Perl IRODS API to read the metadata from the collection, rather than invoking `imeta`. Using the API, you don't need to parse the output yourself with a regex (which is fragile) and you get better test failure messages when testing with `is_deeply` (will pinpoint differences, rather than just failing).